### ISSUE TITLE
FeatureGates in KubeadmControlPlane Spec should be mutable 

### ISSUE THREAD

**Issue by shysank** [kubernetes-sigs/cluster-api#4180]
**User Story**

#2083 was created to make kcp spec fully mutable. But, since it's difficult to track every spec mutation in a single pr, we wanted to start with a few specs that were important, and move from there. As part of #3994 , the following specs in kcp were made mutable:
- APIServer
- ControllerManager
- Scheduler

This story aims to track the `FeatureGates` settings in KCP

**Detailed Description**

Make `FeatureGates` spec in  KubeadmControlPlane mutable

**Anything else you would like to add:**

`v1beta1` in kubeadm has now been deprecated, so we need to keep that in mind when implementing this.

/kind feature

**Comment by fabriziopandini** [comment]
Due to the nature of kubeadm FeatureGates, it is really tricky to make this mutable because there is not a clear/stable definition of what the mutation really impacts on, so you can't really be sure the change will be successfully applied after rollout.

Let's make an example:

- kubeadm introduces FeauteGates X which requires the creation of a new RBAC rule to be generated by kubeadm init.
- change FeatureGates X in KCP
- KCP triggers a rollout, which basically consists of a sequence of kubeadm join commands
However, given that there is no logic in kubeadm join/in KCP to handle the installation of the new RBAC rule, the new feature gate won't be successfully applied.

In other words, given how rollout works today, it is safe to make mutable only a subset of changes to the kubeadm spec, and those changes should be applied by a simple sequence of kubeadm join.

**Comment by vincepri** [comment]
Should we do this, or close the issue?

**Comment by sbueringer** [comment]
With the information Fabrizion provided I would say as there is no safe way to implement it generically I wouldn't implement it right now. 

I think it makes more sense when we run into a concrete use case (with a specific feature gate) and then if there is a need for it implement it for this specific flag?

**Comment by vincepri** [comment]
While I do agree that it might not be safe, I wonder if it's on the same level as the extra flags for components?

**Comment by sbueringer** [comment]
So make them mutable without adding additional implementation for them? 

I think extra flags are applied when recreating the control plane nodes, which is what is expected. There might be a gap between expectation and reality with kubeadm FeatureGates, if I got it correctly that they require execution of `kubeadm init` which we never repeat after the initial cluster creation.


(I assume that we update the ClusterConfiguration and thus new control plane notes pick up the new extra flags)

**Comment by vincepri** [comment]
Yeah that makes sense, I wonder if there is some middle ground because I do see having feature gates might be useful to some, although if we don't plan to have this anytime soon, I'm fine closing this issue.

**Comment by shysank** [comment]
The feature gates only support two flags right now: `PublicKeysECDSA, IPv6DualStack`. Not sure if it is worth to take the risk unless there is an immediate requirement.

**Comment by fabriziopandini** [comment]
Closing this as per discussion above; when can eventually reconsider if a specific uses case arise.
/close

**Comment by sedefsavas** [comment]
Sorry just saw this discussion.
@fabriziopandini 
We need to enable additional flags for enabling CSI (it is a requirement for k8s versions > 1.22) 
For existing clusters to continue to work after upgrading to v1.22, we need to enable these for CAPA:
`CSIMigration + CSIMigrationAWS`, see https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1475 for details.

This may be the case for other providers too @shysank

**Comment by shysank** [comment]
@sedefsavas I think these feature gates are specific to kubeadm. To control the featureGates of kubernetes componenets, you'll have to pass it through extra args. eg:

```
spec:
  kubeadmConfigSpec:
    clusterConfiguration:
      apiServer:
        extraArgs:
          featureGates: CSIMigrationAWS=true
```
These are already allowed in #3994 . @fabriziopandini correct me if I'm wrong.

**Comment by fabriziopandini** [comment]
That's right 
feature gates for the API server (as well for other components) are already mutable; this issue instead was about for kubeadm specific feature flags.
